plugins/redfish/redfish-client.c

```diff
@@ -0,0 +1,598 @@
+		return FALSE;
+	}
+
+	/* this is optional */
+	if (smbios_table != NULL) {
+		g_autoptr(GError) error_smbios = NULL;
+		g_autoptr(GError) error_uefi = NULL;
+		if (!redfish_client_set_smbios_interfaces (self,
+							      smbios_table,
+							      &error_smbios)) {
+			g_debug ("failed to get connection URI automatically: %s",
+				 error_smbios->message);
+		}
+		if (!redfish_client_set_uefi_credentials (self, &error_uefi)) {
+			g_debug ("failed to get username and password automatically: %s",
+				 error_uefi->message);
+		}
+	}
+	if (self->hostname != NULL)
+		g_debug ("Hostname: %s", self->hostname);
+	if (self->port != 0)
+		g_debug ("Port:     %u", self->port);
+	if (self->username != NULL)
+		g_debug ("Username: %s", self->username);
+	if (self->password != NULL)
+		g_debug ("Password: %s", self->password);
+
+	/* try to connect */
+	blob = redfish_client_fetch_data (self, "/redfish/v1", error);
+	if (blob == NULL)
+		return FALSE;
+
+	/* get the update service */
+	if (!json_parser_load_from_data (parser,
+					 g_bytes_get_data (blob, NULL),
+					 (gssize) g_bytes_get_size (blob),
+					 error)) {
+		g_prefix_error (error, "failed to parse node: ");
+		return FALSE;
+	}
+	node_root = json_parser_get_root (parser);
+	if (node_root == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no root node");
+		return FALSE;
+	}
+	obj_root = json_node_get_object (node_root);
+	if (obj_root == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no root object");
+		return FALSE;
+	}
+	g_debug ("Version:  %s",
+		 json_object_get_string_member (obj_root, "ServiceVersion"));
+	g_debug ("UUID:     %s",
+		 json_object_get_string_member (obj_root, "UUID"));
+
+	/* look for UpdateService in Links */
+	if (json_object_has_member (obj_root, "Links"))
+		obj_links = json_object_get_object_member (obj_root, "Links");
+	if (obj_links == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no Links object");
+		return FALSE;
+	}
+	if (json_object_has_member (obj_links, "UpdateService"))
+		obj_update_service = json_object_get_object_member (obj_links, "UpdateService");
+	if (obj_update_service == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "no UpdateService object");
+		return FALSE;
+	}
+	data_id = json_object_get_string_member (obj_update_service, "@odata.id");
+	if (data_id == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no @odata.id string");
+		return FALSE;
+	}
+	self->update_uri_path = g_strdup (data_id);
+	return TRUE;
+}
+
+GPtrArray *
+redfish_client_get_devices (RedfishClient *self)
+{
+	return self->devices;
+}
+
+void
+redfish_client_set_hostname (RedfishClient *self, const gchar *hostname)
+{
+	g_free (self->hostname);
```

Comment:
There's a possibility this is NULL isn't there?

Reply:
I agree most of this time it's going to be NULL. I included it so we could just do multiple calls of `redfish_client_set_hostname()` on the same client from the self tests, although I admit I'm not actually doing that anymore. I think we need to include it, or if not then do a critical/return_val_if_fail or something to avoid silently leaking memory.